### ISSUE TITLE
Sync pending for several days

### ISSUE THREAD

**Issue by rmlockwood** [sillsdev/web-languageforge#150]
I tried to do a sync with my Swedish project and the site said it was queued, but now after several days the site still says "Pending". I'm unable to do anything to change it.

**Comment by megahirt** [comment]
@rmunn this is something we need to look into and make high priority for Monday.  LFMerge is getting 'stuck' and we need to figure out what's wrong, and how we can make the system more reliable.

**Comment by rmunn** [comment]
@megahirt - The immediate cause of the problem seems to be that a previous LfMerge run from August 22 went into an undetected "hanging" state while disposing of the FW project. That process is still running, so I was able to get a stacktrace from it. I couldn't get a complete stacktrace because the C# code on production was compiled without debug symbols, but it looks like a threading deadlock situation: there are two calls to System.Threading.WaitHandle:WaitOne visible in that stacktrace. Full stacktrace below:

```gdb
#0  0x00007fd2ebfdbc1d in nanosleep () at ../sysdeps/unix/syscall-template.S:84
#1  0x000000000060ed71 in _wapi_handle_spin (ms=100) at ../../mono/io-layer/handles-private.h:328
#2  _wapi_handle_timedwait_signal_handle (handle=handle@entry=0x43e, timeout=timeout@entry=0x0, alertable=alertable@entry=1, poll=poll@entry=0) at handles.c:1526
#3  0x000000000060ef3b in _wapi_handle_wait_signal_handle (handle=handle@entry=0x43e, alertable=alertable@entry=1) at handles.c:1493
#4  0x00000000006220db in WaitForSingleObjectEx (handle=0x43e, timeout=timeout@entry=4294967295, alertable=alertable@entry=1) at wait.c:194
#5  0x000000000058e79f in mono_wait_uninterrupted (thread=thread@entry=0x7fd2e96d8010, multiple=multiple@entry=0, numhandles=numhandles@entry=1, handles=handles@entry=0x7ffe63d43438, waitall=waitall@entry=0, ms=ms@entry=-1, alertable=1) at threads.c:1306
#6  0x00000000005903b6 in ves_icall_System_Threading_WaitHandle_WaitOne_internal (this=<optimized out>, handle=0x43e, ms=-1, exitContext=<optimized out>) at threads.c:1438
#7  0x0000000041511488 in ?? ()
#8  0x00000000028ed5d0 in ?? ()
#9  0x00007fd2c2156360 in ?? ()
#10 0x00007fd2c1f90170 in ?? ()
#11 0x00007ffe63d434f0 in ?? ()
#12 0x00007ffe63d43470 in ?? ()
#13 0x00007fd2eb8a3120 in ?? ()
#14 0x00007fd2c1f914b0 in ?? ()
#15 0x00007fd2c1faab58 in ?? ()
#16 0x00007ffe63d435b8 in ?? ()
#17 0x00007fd2e9989f36 in System.Threading.WaitHandle:WaitOne (this=...) at /tmp/buildd/mono-sil-3.4.0/mcs/class/corlib/System.Threading/WaitHandle.cs:709
#18 0x00000000419ab3b6 in ?? ()
#19 0x00007fd2c1faab58 in ?? ()
#20 0x00007fd2bb1237c0 in ?? ()
#21 0x00007fd2e9989f43 in System.Threading.WaitHandle:WaitOne (this=...) at /tmp/buildd/mono-sil-3.4.0/mcs/class/corlib/System.Threading/WaitHandle.cs:709
#22 0x00000000419ab36c in ?? ()
#23 0x00007fd2bb1237c0 in ?? ()
#24 0x00007fd2c1faab58 in ?? ()
#25 0x00007ffe63d435a0 in ?? ()
#26 0x00000000419ab190 in ?? ()
#27 0x00007fd2c7c7d3a8 in ?? ()
#28 0x00007fd2bb1237c0 in ?? ()
#29 0x00007fd2bb19ae00 in ?? ()
#30 0x00007fd2c1f90170 in ?? ()
#31 0x0000000000000000 in ?? ()
```

There are also log entries that prove that the line that hangs is in SynchronizeAction.cs, line 64: `LanguageForgeProject.DisposeFwProject(project);`. There are no WaitOne calls in the LfMerge code that disposes FW projects, so it has to be happening when the Cache object is being disposed. I won't be able to investigate further until I get back to the office on Monday, but in the meantime, I can at least clear the logjam by killing the hung LfMerge process. Then the next one that runs (on the 20-minute cron job) will actually run all the queued S/R processes (just two of them, Ron's Swedish project and the test-rmunn-04 S/R that I just kicked off to verify) and the logjam will be cleared.

**Comment by rmlockwood** [comment]
I’m still unable to do a S/R from LF. It still says pending. Thanks, Ron

**Comment by rmunn** [comment]
@rmlockwood - I just got into the office and unblocked the S/R queue on Language Forge. You should be able to Send/Receive your Swedish project again.

**Comment by megahirt** [comment]
Closing this issue as it has been resolved for the time being.  Please submit another issue if your sync becomes stuck again.
